[vcpkg.cmake] Fix CMP0126 warning when running in trace mode (#23784)

jobor opened the pull request:
When running CMake 3.21 or newer in trace mode, vcpkg.cmake would trigger the CMP0126 warning. Explicitly set this policy to OLD behavior within the vcpkg.cmake policy scope to silence the warning without changing behaviour.

Fixes #23784

jobor commented:
After reviewing the warnings I come to the conclusion that it would be safe to set the policy to NEW within the vcpkg.cmake policy scope. Opinions?

Neumann-A commented:
> Unlike many policies, CMake version 3.26.0-rc5 does not warn when the policy is not set and simply uses OLD behavior. See documentation of the [CMAKE_POLICY_WARNING_CMP0126](https://cmake.org/cmake/help/latest/variable/CMAKE_POLICY_WARNING_CMPNNNN.html#variable:CMAKE_POLICY_WARNING_CMP%3CNNNN%3E) variable to control the warning.

**does not warn when the policy is not set** 

-> CMake bug ? or did you change **CMAKE_POLICY_WARNING_CMPNNNN** ?

dg0yt commented:
> After reviewing the warnings I come to the conclusion that it would be safe to set the policy to NEW within the vcpkg.cmake policy scope.

`vcpkg.cmake` is also used by user projects. So it is not really limited to vcpkg ports. We should be careful before changing it when Kitware decided that it needs a policy.

> **does not warn when the policy is not set**
> 
> -> CMake bug ?

IIUC the behaviours is different with trace modes.

jobor commented:
> -> CMake bug ? or did you change CMAKE_POLICY_WARNING_CMPNNNN ?

In trace mode, CMake enables all optional warnings. This is not a bug.

```
bool cmMakefile::PolicyOptionalWarningEnabled(std::string const& var) const
{
  // Check for an explicit CMAKE_POLICY_WARNING_CMP<NNNN> setting.
  if (cmValue val = this->GetDefinition(var)) {
    return cmIsOn(val);
  }
  // Enable optional policy warnings with --debug-output, --trace,
  // or --trace-expand.
  cmake* cm = this->GetCMakeInstance();
  return cm->GetDebugOutput() || cm->GetTrace();
}
```

> `vcpkg.cmake` is also used by user projects. So it is not really limited to vcpkg ports. We should be careful before changing it when Kitware decided that it needs a policy.

`vcpkg.cmake` uses `cmake_policy(PUSH/POP)`, and within these calls we can set policies without affecting user projects.

Neumann-A commented:
> After reviewing the warnings I come to the conclusion that it would be safe to set the policy to NEW within the vcpkg.cmake policy scope. Opinions?

I don't mind setting it to new.

Neumann-A commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -38,6 +38,10 @@ if(CMAKE_VERSION VERSION_LESS Z_VCPKG_CMAKE_REQUIRED_MINIMUM_VERSION)
 endif()
 cmake_policy(PUSH)
 cmake_policy(VERSION 3.7.2)
+if(CMAKE_VERSION VERSION_GREATER_EQUAL "3.21.0")
```
```suggestion
if(POLICY CMP0126)
```

Neumann-A commented:
> In trace mode, CMake enables all optional warnings. This is not a bug.

But the code also says you could actively suppress the warning if I read it correctly. (i.e. by explicitly setting it to false)

dg0yt commented:
Trying to get the full picture.
- The code run at top-level scope.
- It is scoped by `cmake_policy(PUSH/POP)`.
  However, there are two ways how this policy setting enters the scope of the user projects:
  - The chainloaded user toolchain file is loaded in the same scope.
  - The policy setting is used by vcpkg macros which override cmake commands (#28960).

I'm not really against this change. But I want all options to be considered:
- Do nothing. The user can adjust policies and warnings if desired. Standard behaviour.
- Force-disable the warning. Needs care to not override an explicit user setting. Mild non-standard behaviour.
- Force-enable the policy. This PR. Recommended for new CMake code, but would also apply on old find modules and package configs.

Last not least it might make sense to actually check the current warnings. In #23784, I see `VCPKG_MANIFEST_DIR`. This is a variable under vcpkg's control. Can we change our scripts to avoid the warning?

jobor commented:
Avoiding the warning without modifying the policies is certainly an option. The pattern where the warning is triggered is:
```
if(NOT VCPKG_SOME_VAR)
    set(VCPKG_SOME_VAR some default)
endif()
set(VCPKG_SOME_VAR some default CACHE PATH "docstring" FORCE)
```

The minimal impact fix is
```
if(NOT VCPKG_SOME_VAR)
    set(VCPKG_SOME_VAR some default CACHE PATH "docstring")
endif()
set(VCPKG_SOME_VAR some default CACHE PATH "docstring" FORCE)
```

Or it might be more elegant to use a differently named variable for holding the default value.

If I remember correctly, only two VCPKG_* variables are affected. 
I'll play with this approach.

jobor committed:
[vcpkg.cmake] Fix CMP0126 warnings (#23784)

When running CMake 3.21 or newer in trace or debug mode, vcpkg.cmake
would trigger the CMP0126 warning for VCPKG_MANIFEST_DIR and
VCPKG_INSTALLED_DIR.

The regular variable was used to set the initial value of the cache
variable of the same name. This patch adds the regular variables
Z_VCPKG_MANIFEST_DIR_INITIAL_VALUE and
Z_VCPKG_INSTALLED_DIR_INITIAL_VALUE that are used to initialize their
respective cache variables.

FrankXie05 commented:
It's better to adapt to warnings than to avoid them, and I think this one is ok.
We can check whether this warning is triggered elsewhere, and we can refer to the modification.  :)
